app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt

```diff
@@ -0,0 +1,83 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.toolbar
+
+import android.content.Context
+import android.view.View
+import android.widget.TextView
+import androidx.annotation.ColorRes
+import androidx.annotation.DrawableRes
+import androidx.core.content.ContextCompat
+import mozilla.components.browser.menu.BrowserMenu
+import mozilla.components.browser.menu.item.BrowserMenuImageText
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+
+internal const val NO_ID = -1
+
+internal fun TextView.setColorResource(@ColorRes textColorResource: Int) {
+    if (textColorResource != NO_ID) {
+        setTextColor(ContextCompat.getColor(context, textColorResource))
+    }
+}
+
+/**
+ * A helper function to get the proper label for the sign in button.  There are three states:
+ * user is signed-in, please sign-in, and we are not sure what's happening. The last state does
+ * occur when the menu is opened before the AccountManager has finished updating its state, or
+ * when there is no connection and AccountManager doesn't return us the profile because its cache
+ * has 2 minutes 'freshness':
+ * https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41
+ */
+fun Context.getSignInLabel(): String {
+    val isAuthenticated = components.backgroundServices.accountManager.authenticatedAccount() != null
+    val isAccountManagerReady = components.backgroundServices.accountManagerAvailableQueue.isReady()
+    val email = components.backgroundServices.accountManager.accountProfile()?.email
+
+    return if (isAccountManagerReady) {
+        if (isAuthenticated) {
+            // It sounds a bit counterintuitive, but authenticated state doesn't mean we
+            // have the profile information. The state can be (and usually is) restored
+            // without connection to the Internet, but profile can not. So we have to make
+            // sure that we actually have the email in order to show it.
+            email ?: resources.getString(R.string.browser_menu_account_settings)
+        } else {
+            getString(R.string.sync_menu_sync_and_save_data)
+        }
+    } else {
+        resources.getString(R.string.browser_menu_account_settings)
+    }
+}
+
+/**
+ * A menu item for displaying account information. The item calculates the label on every bind call,
+ * to provide each menu with the latest account manager state.
+ *
+ * @param label The function that provides the visible label of this menu item.
+ * @param imageResource ID of a drawable resource to be shown as icon.
+ * @param textColorResource Optional ID of color resource to tint the text.
+ * @param listener Callback to be invoked when this menu item is clicked.
+ */
+
+class BrowserMenuSignIn(
+    private val label: () -> String,
+    @DrawableRes
+    imageResource: Int,
+    @ColorRes
+    private val textColorResource: Int,
```

Comment:
nit: if you keep the class, the annotations could be moved onto the same line as the parameter name for better readability